subprojects/base-services/src/main/java/org/gradle/internal/service/DefaultServiceRegistry.java

```diff
@@ -296,23 +302,15 @@ private static String format(Type type) {
         return type.toString();
     }
 
-    public boolean hasService(Class<?> serviceType) {
-        assertValidServiceType(serviceType);
-        return allServices.hasService(serviceType);
-    }
-
     public <T> List<T> getAll(Class<T> serviceType) throws ServiceLookupException {
         synchronized (lock) {
             mutable = false;
             if (closed) {
                 throw new IllegalStateException(String.format("Cannot locate service of type %s, as %s has been closed.", format(serviceType), getDisplayName()));
             }
-            if (!hasService(serviceType)) {
-                return Collections.emptyList();
-            }
+            assertValidServiceType(serviceType);
             List<ServiceProvider> providers = new ArrayList<ServiceProvider>();
-            DefaultLookupContext context = new DefaultLookupContext();
-            allServices.getAll(context, serviceType, providers);
+            collectInto(serviceType, providers);
```

Comment:
:o: we could probably avoid the double `ArrayList<>` creation, by having a collect method that would directly work on `List<T>` (and call `getProvider()` for each collected service).

Reply:
The main reason we use `ServiceProvider` on the `getAll` API is so we can track dependencies between services. It might be possible to get rid of that, but it doesn't seem straightforward.

Reply:
I gave this another look and you were right, this was easier than I thought - we now only have one array and a wrapping list.